gravitee-am-factor/gravitee-am-factor-email/src/main/java/io/gravitee/am/factor/email/provider/EmailFactorProvider.java

```diff
@@ -168,17 +143,45 @@ public Completable sendChallenge(FactorContext context) {
     private Completable generateCodeAndSendEmail(FactorContext context, EmailSenderProvider provider, EnrolledFactor enrolledFactor) {
         logger.debug("Generating factor code of {} digits", configuration.getReturnDigits());
 
-
         try {
-            final String recipient = enrolledFactor.getChannel().getTarget();
+            UserService userService = context.getComponent(UserService.class);
+
+            EmailService emailService = context.getComponent(EmailService.class);
+            Email mfaTemplate = emailService.getEmailTemplate(Template.MFA_CHALLENGE, context.getClient());
+
+            Configuration freemarkerConfiguration = context.getComponent(Configuration.class);
+
+            Map<String, Object> params =  context.getTemplateValues();
             // register mfa code to make it available into the TemplateEngine values
-            context.registerData(FactorContext.KEY_CODE, generateOTP(enrolledFactor));
-            // Generate template before creating the code into repositories to avoid useless entries in case of template error
-            TemplateEngine templateEngine = context.getTemplateEngine();
-            String content = templateEngine.getValue(configuration.getTemplate(), String.class);
-            String subject = templateEngine.getValue(configuration.getSubject(), String.class);
+            params.put(FactorContext.KEY_CODE, generateOTP(enrolledFactor));
+            params.put("expireAfterSeconds", mfaTemplate.getExpiresAfter());
+            final long expiresAt = Instant.now().plus(mfaTemplate.getExpiresAfter(), ChronoUnit.SECONDS).toEpochMilli();
+            params.put("expireAt", expiresAt);
+
+            // compute email subject
+            final freemarker.template.Template plainTextTemplate = new freemarker.template.Template("subject", new StringReader(mfaTemplate.getSubject()), freemarkerConfiguration);
+            final String subject = processTemplateIntoString(plainTextTemplate, params);
+
+            // compute email content
+            final freemarker.template.Template template = freemarkerConfiguration.getTemplate(mfaTemplate.getTemplate().endsWith(TEMPLATE_SUFFIX) ? mfaTemplate.getTemplate() : mfaTemplate.getTemplate() + TEMPLATE_SUFFIX);
+            final String content = processTemplateIntoString(template, params);
 
-            return provider.sendMessage(new Message(recipient, content, subject, configuration.getContentType()));
+            final String recipient = enrolledFactor.getChannel().getTarget();
+
+            io.gravitee.am.common.email.Email email = new EmailBuilder()
+                    .from(mfaTemplate.getFrom())
+                    .fromName(mfaTemplate.getFromName())
+                    .subject(subject)
+                    .to(recipient)
+                    .build();
+            email.setContent(content);
```

Comment:
I wonder if we cannot do all this technical stuff via the emailService ... something like : 

`Single<Email> emailService.createEmailTemplate(template, params, client);`

This way we remove all our internal templating mechanism (freemarker, ...)